### PR TITLE
Wandb, cpu offloading, cosine scheduler, CodeLlama, Completion and bug fixes

### PR DESCRIPTION
**Pull request by maximegmd**
# What does this PR do?

This PR contains a bit of everything I did (and will do) to improve my workflow, it can be merged but I am opening it with the intent of making these changes visible to users who are looking for them.

- Add FSDP CPU offloading for smaller configs #122 
- Add support for wandb tracking
- Add support for cosine + warmup scheduler
- Add support for CodeLlama
- Add support for completion corpus datasets with optional packing or padding

Fixes:
- Progress bar not incrementing correctly and displaying erroneous numbers.

## Feature/Issue validation/testing

- [x] LLama 7b full train on 4xV100 16GB
- [x] Llama 7b PEFT Lora on 4xV100 16GB
- [x] CodeLlama 7b full train on 4xV100 16GB

### CODE REVIEW

File: requirements.txt [hunk 1]
```diff
@@ -9,8 +9,9 @@ black[jupyter]
 datasets
 fire
 git+https://github.com/huggingface/peft.git
-transformers>=4.31.0
+git+https://github.com/huggingface/transformers.git@main
```
Comment (mreso) [inline requirements.txt:12]: Why are we going from release pypi to rolling release github here?

File: utils/train_utils.py [hunk 2]
```diff
@@ -45,7 +46,7 @@ def set_tokenizer_params(tokenizer: LlamaTokenizer):
 def byte2mb(x):
     return int(x / 2**20)
 
-def train(model, train_dataloader,eval_dataloader, tokenizer, optimizer, lr_scheduler, gradient_accumulation_steps, train_config, fsdp_config=None, local_rank=None, rank=None):
+def train(model, train_dataloader,eval_dataloader, tokenizer, optimizer, lr_scheduler, gradient_accumulation_steps, train_config, fsdp_config=None, local_rank=None, rank=None, wandb_run=None, step_scheduler=False):
```
Comment (mreso) [inline utils/train_utils.py:49]: Do we need an extra parameter for step_scheduler? Can this not be lr_scheduler and we thread is differently based on type (if necessary)?

File: utils/train_utils.py [hunk 3]
```diff
@@ -101,16 +104,25 @@ def train(model, train_dataloader,eval_dataloader, tokenizer, optimizer, lr_sche
...
+                            lr_scheduler.step()
                 else:
                     # regular backpropagation when fp16 is not used
                     loss.backward()
                     if (step + 1) % gradient_accumulation_steps == 0 or step == len(train_dataloader) - 1:
                         optimizer.step()
                         optimizer.zero_grad()
-                        pbar.update(step//gradient_accumulation_steps)
+                        pbar.update()
+                        if step_scheduler:
```
Comment (mreso) [inline utils/train_utils.py:118]: Is it relevant to update scheduler on different places? Would one not be sufficient and you just adapt the step length? Like we call update every step instead of every second but double step size?

File: utils/train_utils.py [hunk 4]
```diff
@@ -396,8 +412,6 @@ def save_train_params(train_config, fsdp_config, rank):
     train_config.dist_checkpoint_root_folder
     + "/"
     + train_config.dist_checkpoint_folder
-    + "-"
```
Comment (mreso) [inline utils/train_utils.py:399]: Why this change?

File: configs/training.py [hunk 5]
```diff
@@ -30,6 +30,8 @@ class train_config:
     quantization: bool = False
     one_gpu: bool = False
     save_model: bool = True
+    use_cosine_scheduler: bool = False
```
Comment (mreso) [inline configs/training.py:33]: There are many different schedulers, can we make this an easily extensible enum or similar to avoid having n use_X_scheduler bools? Given that all need different parameters a separate config dataclass for schedulers would work as well and fit our pattern. Compare datasets

File: configs/wandb.py [hunk 6]
```diff
@@ -0,0 +1,16 @@
...
+from dataclasses import dataclass, field
+
+@dataclass
+class wandb_config:
+    enable_wandb: bool = False
+    wandb_id: str = None
+    wandb_project: str = None
+    wandb_entity: str = None
+    wandb_mode: str = None
+    wandb_dir: str = None
```
Comment (mreso) [inline configs/wandb.py:13]: Prefix is not necessary as its inside wandb_config. You will configure it through --wandb_config.dir from the command line instead of --wandb_config.wandb_dir. Could we populate some of these (like dir) with meaningful default values?

File: configs/wandb.py [hunk 7]
```diff
@@ -0,0 +1,16 @@
+# Copyright (c) Meta Platforms, Inc. and affiliates.
+# This software may be used and distributed according to the terms of the Llama 2 Community License Agreement.
+
+from dataclasses import dataclass, field
+
+@dataclass
+class wandb_config:
+    enable_wandb: bool = False
```
Comment (mreso) [inline configs/wandb.py:8]: We should move enable wandb in training_config and only then read in this config.

File: ft_datasets/__init__.py [hunk 8]
```diff
@@ -3,4 +3,5 @@
 
 from .grammar_dataset import get_dataset as get_grammar_dataset
 from .alpaca_dataset import InstructionDataset as get_alpaca_dataset
-from .samsum_dataset import get_preprocessed_samsum as get_samsum_dataset
\ No newline at end of file
+from .samsum_dataset import get_preprocessed_samsum as get_samsum_dataset
+from .completion_dataset import get_preprocessed_completion as get_completion_dataset
```
Comment (mreso) [inline ft_datasets/__init__.py:7]: Can we rename completion to code completion to make it more obvious to the user?

File: llama_finetuning.py [hunk 9]
```diff
@@ -12,21 +12,23 @@
 from torch.distributed.fsdp import (
     FullyShardedDataParallel as FSDP,
 )
+from torch.distributed.fsdp.fully_sharded_data_parallel import CPUOffload
```
Comment (mreso) [inline llama_finetuning.py:15]: You can import CPUOffload directly from torch.distributed.fsdp together in line 12

File: llama_finetuning.py [hunk 10]
```diff
@@ -12,21 +12,23 @@
...
 from torch.optim.lr_scheduler import StepLR
 from torch.utils.data import DistributedSampler
 from transformers import (
     LlamaForCausalLM,
-    LlamaTokenizer,
+    AutoTokenizer,
```
Comment (mreso) [inline llama_finetuning.py:20]: Why do we change this to AutoTokenizer?

File: utils/wandb_utils.py [hunk 11]
```diff
@@ -0,0 +1,16 @@
+import os
+from configs import wandb_config
+
+def wandb_watch(model, cfg: wandb_config):
+    import wandb
```
Comment (mreso) [inline utils/wandb_utils.py:5]: As we're importing wandb only if we enable it we could potentially make it an optional dependency (compare vllm in pyproject.toml). Though I am not sure if its not installed as a dependency already.

File: ft_datasets/completion_dataset.py [hunk 12]
```diff
@@ -0,0 +1,31 @@
+# Copyright (c) Meta Platforms, Inc. and affiliates.
```
Comment (mreso) [inline ft_datasets/completion_dataset.py:1]: Would be good to integrate a test for the dataset.

File: utils/train_utils.py [hunk 13]
```diff
@@ -101,16 +104,25 @@ def train(model, train_dataloader,eval_dataloader, tokenizer, optimizer, lr_sche
                         scaler.step(optimizer)
                         scaler.update()
                         optimizer.zero_grad()
-                        pbar.update(step//gradient_accumulation_steps)
+                        pbar.update()
+                        if step_scheduler:
```
Comment (mreso) [inline utils/train_utils.py:108]: see below

**Review by mreso — CHANGES_REQUESTED** [review]
Thanks for the contribution! Left a few comments. Please also address the conflicts.

File: requirements.txt (hunk 1 above)
Comment (maximegmd) [reply requirements.txt:12]: The fast code llama tokenizer wasn't working correctly and only the git version had the fix at the time of this change, hopefully it can be rolled back now

File: utils/train_utils.py (hunk 4 above)
Comment (maximegmd) [reply utils/train_utils.py:399]: When passing full paths as the model's name it creates chaotic paths

File: ft_datasets/__init__.py (hunk 8 above)
Comment (maximegmd) [reply ft_datasets/__init__.py:7]: it's not code completion, it's just corpus text completion, it's what I use for pretraining from textbooks for example

File: llama_finetuning.py (hunk 10 above)
Comment (maximegmd) [reply llama_finetuning.py:20]: To support the CodeLLamaTokenizer automatically
